### DISCUSSION TITLE
Setup notes for Ubuntu 22.04 with A4000 GPU

### DISCUSSION THREAD

**Discussion by bbrowning** [instructlab/instructlab#660] [Show and tell]
I didn't have a local machine powerful enough to run `lab train`, so after a few iterations settled on an A4000 GPU running Ubuntu 2204 from Paperspace as a temporary and cheaper solution that buying new personal hardware. Here are some pretty raw notes from how I got that setup and working, in case anyone else finds themselves in a similar Ubuntu 22.04 + Ampere situation and trying to get things going.

# PaperSpace A4000 GPU w/ Ubuntu 22.04

Ensure the machine has at least 100GB of storage or more - I chose 250GB. You pay a small rate 24x7 for the storage, but the main cost comes from the compute which only gets charged while the VM is running.

## Update Machine

- `sudo apt update`
- `sudo apt upgrade`
- `sudo reboot`

## Install Python 3.11
- `sudo apt update` (sometimes apt cache seems empty after a reboot but updated fixes it)
- `sudo apt install build-essential zlib1g-dev libncurses5-dev libgdbm-dev libnss3-dev libssl-dev libreadline-dev libffi-dev libsqlite3-dev wget libbz2-dev liblzma-dev`
- `mkdir src`
- `cd src`
- `wget https://www.python.org/ftp/python/3.11.8/Python-3.11.8.tgz`
- `tar -xvf Python-3.11.8.tgz`
- `cd Python-3.11.8/`
- `./configure --enable-optimizations`
- `make -j 8`
- `sudo make altinstall` (puts python3.11 in /usr/local/bin)
- `python3.11 --version`

## Clone InstructLab repos
- `ssh-keygen` and add that new key to the github.com via its UI
- `cd ~/src`
- clone instruct-lab/taxonomy repo (or your fork of it) locally
- clone instruct-lab/cli repo (or your fork of it) locally
- alternatively, use rsync over ssh to copy the repos from your local machine to the cloud VM

## Setup CLI repo virtualenv
- cd into your instruct-lab cli directory
- `python3.11 -m venv venv`
- `source venv/bin/activate`

## Checkout Linux GPU offload PR #520
- `git fetch origin pull/520/head:linux-gpu`
- `git switch linux-gpu`
- `pip3 install -e .`

## Follow GPU Acceleration docs in CLI repo
- read docs/gpu-acceleration.md and follow those instructions
- what I did below:

### Install CUDA and Nvidia Proprietary Drivers
- `sudo apt install linux-headers-$(uname -r)`
- `wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64/cuda-keyring_1.1-1_all.deb`
- `sudo dpkg -i cuda-keyring_1.1-1_all.deb`
- `sudo apt update`
- `sudo apt -y install cuda-toolkit-12-4`
- `sudo apt install -y cuda-drivers`
  - this installs the Nvidia driver that matches your cuda version - don't make the mistake I originally did and install a different way
- `sudo apt install ubuntu-drivers-common`
- `sudo apt install nvidia-utils-535-server`
  - this should technically match the drive installed via cuda-drivers, which was 550 I think? It took me a few tries with driver installation to get things working, but 535 utils with the 550 driver seemed ok
- `sudo reboot`
  - wait for machine to reboot, ensure it comes back up with nvidia driver in use
- `lspci -n -n -k | grep -A 2 -e VGA -e 3D`
  - make sure that outputs nvidia as the driver

### Replace llama-cpp-python with cuda version
- cd into your instruct-lab cli directory
- ` source venv/bin/activate`
- `export CUDA_HOME=/usr/local/cuda`
- `export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64`
- `export PATH=$PATH:$CUDA_HOME/bin`
- `CMAKE_ARGS="-DLLAMA_CUBLAS=on" pip3 install --force-reinstall --no-cache-dir llama-cpp-python`
- `pip3 install -e .`

## Test lab cli
- `lab init` (give it the path to your taxonomy repo)
- `lab download`
- `lab serve`
- `lab chat` (in a new terminal, within your python venv) 
- at this point, everything should be working - you can confirm GPU
  usage with `nvidia-smi -l` in another terminal while chatting

## Generate Examples
- make any small modification (whitespace, comment) to an existing
  qna.yaml in the taxonomy repo
- `lab generate`
  - INFO 2024-03-16 12:49:10,610 generate_data.py:439 Generation took 443.38s

## Train
- You'll need to make some local modifications to `cli/train/linux_train.py` - see Ben's changes shown in https://github.com/instruct-lab/cli/pull/520#issuecomment-2001986101 for what worked for him on this 16GB A4000 GPU
- `lab train --device cuda`
  - this should take about 25 minute or so if you did everything correctly

**Comment by tiran** [comment]
Two suggestions:

## Install Python 3.11

Instead of building Python from sources manually, you either use Deadsnakes PPA or https://github.com/pyenv/pyenv . Deadsnakes PPA provides Debian packages for older or more recent Python versions. pyenv automates building from source.

## Clone InstructLab repos

Please don't suggest to copy SSH keys to other machines. Private SSH keys should not leave your machine. I recommend that you either copy the cli and taxonomy directory (rsync over ssh works great) or use [SSH agent forwarding](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/using-ssh-agent-forwarding) (`ssh -A`). Agent forwarding is less secure, but still much better than copying key file to an untrusted cloud machine.

**Reply by bbrowning** [reply]
Both good points - I created new SSH keys on this machine just for this purpose, so will edit it to suggest that or copying the files over via rsync over ssh. And I'll look at the other Python options, and update instructions once I've tried one of them.

**Comment by bbrowning** [comment]
Closing as outdated, given guidance that this shouldn't be done until the project is public.
